rep-0147.rst

```diff
@@ -0,0 +1,365 @@
+
+
+Reference Abstractions
+======================
+
+For this document we will use the following abstractions of the architecture.
+These abstractions are used to define the interfaces, however it is not required that a given system implement all the interfaces.
+A given system may implement subsets of these interfaces.
+
+
+Control Abstractions
+--------------------
+
+For controlling an aerial vehicle we will use the following terms for the interfaces.
+
+.. raw:: html
+  
+  <div class="mermaid">
+  graph LR
+    na[ ]
+    tc[Trajectory Controller]
+    ac[Attitude Controller]
+    rc[Rate Controller]
+    va[Vehicle Abstraction]
+    na ==>|Trajectory|tc
+    tc ==>|Pose|ac
+    ac ==>|Rate|rc
+    rc ==>|Acceleration|va
+    
+    va -.-> am[Allocation Matrix]
+    va -.-> mixer
+    
+    am -.-> Actuators
+    mixer -.-> Actuators 
+  </div>
+
+High Level Architecture
+-----------------------
+
+A higher level view of the system uses the following abstraction.
+
+.. raw:: html
+  
+  <div class="mermaid">
+  graph LR
+    GCS[Ground Control Station]
+    CS[Collision Sensors]
+    OS[Odometry Sensors]
+    C[Controller]
+    L[ ]
+    style L fill:#fff,stroke:#ff0,stroke-width:0px;
+    L --> CS
+    L --> GCS
+    L --> OS
+    subgraph Generic Planning Framework
+    CM[Collision Mapping]
+    MSC[Minimum Snap Controller]
+    CM ---|Shared State|MSC
+    end
+    CS -->|Obstacle Observations|CM
+    OS -->|Odometry|MSC
+    OS -->|Odometry|CM
+    OS -->|Odometry|C
+    GCS -->|Goals|MSC
+    MSC -->|Trajectory|C
+    %% Links to L 
+    linkStyle 0 fill:#fff,stroke:#ff0,stroke-width:0px;
+    linkStyle 1 fill:#fff,stroke:#ff0,stroke-width:0px;
+    linkStyle 2 fill:#fff,stroke:#ff0,stroke-width:0px;
+  </div>
+
+
+Coordinate Frames
+=================
+
+It is expected that users of this interface will also follow the coordinate frame conventions layed out in REP 105 [#rep105]_.
+
+Proposed Interfaces
+===================
+
+Control Interfaces
+------------------
+
+Trajectory Control Interface
+,,,,,,,,,,,,,,,,,,,,,,,,,,,,
+
+For sending trajectories in 3D space there already exists a message used for free space planning.
+This message used is ``trajectory_msgs/MultiDOFJointTrajectory`` [#multidoftraj]_ and is actively used by planning frameworks already.
+
+This message is not the most straight forward to parse as it has a lot of optional fields which increases the complexity and requires consistent labelling and filling.
+The complexity of the message parsing is traded off against supporting a relatively compact representations.
+
+* Datatype: ``trajectory_msgs/MultiDOFJointTrajectory`` [#multidoftraj]_ 
+* Topic: ``command_trajectory``
+
+Pose Interface
+,,,,,,,,,,,,,,
+
+The message to send a goal pose is ``geometry_msgs/PoseStamped`` [#posestamped]_.
+This is a very highly used standard message and contains precisely the information needed.
+
+* Datatype: ``geometry_msgs/PoseStamped`` [#posestamped]_
+* Topic: ``command_pose``
```

Comment:
1. Would it be possible to command only a subset of the components in `geometry_msgs/PoseStamped` without affecting the others? For example, the value NaN might have the meaning "don't change".

   Examples:
   - go to position x/y at the same altitude:
      ```
      pose:
        position:
          x: 10.0
          y: 20.0
          z: NaN
        orientation:
          x: NaN
          y: NaN
          z: NaN
          w: NaN
      ```

   - climb/descend to altitude z:
      ```
      pose:
        position:
          x: NaN
          y: NaN
          z: 5.0
        orientation:
          x: NaN
          y: NaN
          z: NaN
          w: NaN
      ```

   - change heading to 180 (nose pointing to the south) without changing position or altitude:
      ```
      pose:
        position:
          x: NaN
          y: NaN
          z: NaN
        orientation:
          x: 0.0
          y: 0.0
          z: 1.0
          w: 0.0
      ```
     A vehicle might ignore certain components of the orientation if it is not able to control them independently and in   a stable manner.

2. Does this interface also allow relative position changes, like "fly 2 meters to the north"? The `frame_id` could be used to encode such a command, but then we would need two new frames centered at the vehicle, which are also interesting for a velocity interface (see [comment above](https://github.com/ros-infrastructure/rep/pull/118/files#r68248878)):

   - `body_enu`: centered at the vehicle (like body), but axes pointing to ENU
   - `body_xy`: centered at the vehicle, but x axis being parallel to the ground and pointing towards the heading of the vehicle and z axis pointing upwards

   Is there already a definition of those frames somewhere in ROS?

3. What would be the meaning of the header timestamp? E.g.:

   - 0.0 = execute immediately
   - start moving at ... or
   - be there at ... (implicit velocity information)
   - transform frame given in the command at time t and execute immediately, e.g. if the command was given in body frame